docs/rfcs/033-ancestor-deletion.md

```diff
@@ -0,0 +1,237 @@
+other timelines that depended on the parent. If we detached many children this
+way, the storage cost would become prohibitive (consider a 1TB database with
+100 child timelines: it would cost 100TiB if they all generated their own image layers).
+
+### Don't rewrite anything: just fake it in the API
+
+We could add a layer of indirection that let a child "pretend" that it had no
+ancestor, when in reality it still had the parent. The pageserver API could
+accept deletion of ancestor timelines, and just update child metadata to make
+them look like they have no ancestor.
+
+This would not achieve the desired reduction in storage cost, and may well be more
+complex to maintain than simply implementing the API described in this RFC.
+
+### Avoid copying objects: enable child index to use parent layers directly
+
+We could teach IndexPart to store a TimelineId for each layer, such that a child
+timeline could reference a parent's layers directly, rather than copying them
+into the child's prefix.
+
+This would impose a cost for the normal case of indices that only target the
+timeline's own layers, add complexity, and break the useful simplifying
+invariant that timelines "own" their own path. If child timelines were
+referencing layers from the parent, we would have to ensure that the parent
+never runs GC/compaction again, which would make the API less flexible (the
+proposal in this RFC enables deletion of the parent but doesn't require it.)
+
+## Performance
+
+### Adopting layers
+
+- CopyObject is a relatively cheap operation, but we may need to issue tens of thousands
+  of such requests: this can take up to tens of seconds and will compete for RemoteStorage
+  semaphore units with other activity on the pageserver.
+- If we are running on storage backend that doesn't implement CopyObject, then
+  this part will be much more expensive as we would stream all layer content
+  through the pageserver. This is no different to issuing a lot
+  of reads to a timeline that does not have a warm local cache: it will move
+  a lot of gigabytes, but that shouldn't break anything.
+- Generating truncated layers for delta that straddle the branch point will
+  require streaming read/write of all the layers in question.
+
+### Updating timeline ancestry
+
+The simplest way to update timeline ancestry will probably be to stop and start
+all the Timeline objects: this is preferable to the complexity of making their
+ancestry mutable at runtime.
+
+There will be a corresponding "stutter" in the availability of the timelines,
+of the order 10-100ms, which is the time taken to upload their IndexPart, and
+restart the Timeline.
+
+# Interaction with other features
+
+## Concurrent timeline creation
+
+If new historic timelines are created using the parent as an ancestor while the merge API is running, they will not be re-parented to the child. This doesn't break
+anything, but it leave the parent in a state where it might not be possible
+to delete it.
```

Comment:
Inability to delete is a breakage from the end user point of view, as it will likely result in the tenant becoming stuck because the deletion of the old parent failed. Below you suggest the following:

> they can choose to delete those children to enable deleting the parent.

We likely do not want to delete that timeline because it might potentially have had writes.

I'd also like some discussion about race conditions, say if timeline creation is spawned in the background (currently that doesn't happen but it might in the future) and it isn't finished yet by the time the merge API endpoint is called, so it doesn't appear as child for the merge.

For example the RFC could say that callers of the pageserver APIs need to ensure that all creations have finished before a merge is conducted.

Reply:
This is a rather interesting dimension, as in, should there be a reparenting API as well. Theoretically we could do reparenting by hand, but in practice not that easily.

> For example the RFC could say that callers of the pageserver APIs need to ensure that all creations have finished before a merge is conducted.

This sounds good for me.

Reply:
[39054a6](https://github.com/neondatabase/neon/pull/6888/commits/39054a6191eb3a5170f2609a0b3dbb0d2fb62058)